### PR TITLE
feat(overlay): make overlays synchronous

### PR DESCRIPTION
**Pull request by jelbourn**
R: @kara @hansl 

Also portals. 
This dramatically cleans up a lot of stuff and is probably faster.

If I was smart I would have merged #1041 first.

### CODE REVIEW

**Comment by jelbourn** [comment]
This might also fix our unit test flakiness...

File: src/lib/dialog/dialog.ts [hunk 1]
```diff
@@ -39,21 +39,21 @@ export class MdDialog {
...
-    let overlayRef: OverlayRef;
+  open<T>(component: ComponentType<T>, config: MdDialogConfig): MdDialogRef<T> {
+    let overlayRef = this._createOverlay(config);
+    let dialogContainer = this._attachDialogContainer(overlayRef, config);
 
-    return this._createOverlay(config)
-        .then(overlay => overlayRef = overlay)
-        .then(overlay => this._attachDialogContainer(overlay, config))
-        .then(containerRef => this._attachDialogContent(component, containerRef, overlayRef));
+    // TODO: probably need to wait for dialogContainer ngOnInit before attaching.
```
Comment (hansl) [inline src/lib/dialog/dialog.ts]: This seems to imply we should still by async...?

**Comment by hansl** [comment]
One question, but LGTM in general.

File: src/lib/dialog/dialog.ts (hunk 1 above)
Comment (jelbourn) [inline src/lib/dialog/dialog.ts]: Was an in-progress thought while I was working, ended up not being the case. Removed.

**Pull request merged**